zookeeper/src/main/java/com/yahoo/ycsb/db/zookeeper/ZKClient.java

```diff
@@ -0,0 +1,207 @@
+      watcher = null;
+    } else if ("true".equals(watchFlagString)) {
+      watcher = new SimpleWatcher();
+    } else {
+      throw new IllegalArgumentException("zookeeper.watchFlag only has two options:true or false");
+    }
+
+    long sessionTimeout;
+    String sessionTimeoutString = props.getProperty(SESSION_TIMEOUT_PROPERTY);
+    if (sessionTimeoutString != null) {
+      sessionTimeout = Integer.parseInt(sessionTimeoutString);
+    } else {
+      sessionTimeout = DEFAULT_SESSION_TIMEOUT;
+    }
+
+    try {
+      zk = new ZooKeeper(connectString, (int) sessionTimeout, null);
+    } catch (IOException e) {
+      throw new DBException("Creating connection failed.");
+    }
+  }
+
+  public void cleanup() throws DBException {
+    try {
+      zk.close();
+    } catch (InterruptedException e) {
+      throw new DBException("Closing connection failed.");
+    }
+  }
+
+  @Override
+  public Status read(String table, String key, Set<String> fields,
+                     Map<String, ByteIterator> result) {
+    String path = getPath(key);
+    try {
+      byte[] data = zk.getData(path, watcher, null);
+      if (data == null || data.length == 0) {
+        return Status.OK;
+      }
+      JSONObject jsonObject = (JSONObject)JSONValue.parse(new String(data));
+      Iterator<String> iterator = jsonObject.keySet().iterator();
+      while(iterator.hasNext()) {
+        String field = iterator.next();
+        String value = jsonObject.get(field).toString();
+        result.put(field, new StringByteIterator(value));
+      }
+      return Status.OK;
+    } catch (KeeperException | InterruptedException e) {
+      LOG.error("Error when reading a path:{},tableName:{}", path, table, e);
+      return Status.ERROR;
+    }
+  }
+
+  @Override
+  public Status insert(String table, String key,
+                       Map<String, ByteIterator> values) {
+    String path = getPath(key);
+    String data = getJsonStrFromByteMap(values);
+    try {
+      zk.create(path, data.getBytes(UTF_8), ZooDefs.Ids.OPEN_ACL_UNSAFE,
+          CreateMode.PERSISTENT);
+      return Status.OK;
+    } catch (KeeperException.NodeExistsException e1) {
+      return Status.OK;
+    } catch (KeeperException | InterruptedException e2) {
+      e2.printStackTrace();
+      return Status.ERROR;
+    }
+  }
+
+  @Override
+  public Status delete(String table, String key) {
+
+    String path = getPath(key);
+    try {
+      zk.delete(path, -1);
+      return Status.OK;
+    } catch (InterruptedException | KeeperException e) {
+      LOG.error("Error when deleting a path:{},tableName:{}", path, table, e);
+      return Status.ERROR;
+    }
+  }
+
+  @Override
+  public Status update(String table, String key,
+                       Map<String, ByteIterator> values) {
+    String path = getPath(key);
+    String data = getJsonStrFromByteMap(values);
+    try {
+      zk.setData(path, data.getBytes(UTF_8), -1);
```

Comment:
Unless I'm missing something, I think this is overwriting any fields that aren't in the update to be empty?

Reply:
- zookeeper does support the simple k-v data struct which the value is plaintext, unlike the other no-sql database(e.g redis,hbase).
- For the update, the value can be updated from multi field-values to only one field-value.For the benchmark,It doesn't matter?

Reply:
it matters because you aren't doing the action the benchmark is trying to measure. In the current implementation you have to pay the cost of reading and deserializing the current record in order to do an update of a subset of fields.

Reply:
It's tricky.

> you have to pay the cost of reading and deserializing the current record in order to do an update of a subset of fields.

Yep.Since the read operation only reads from the memory, so the cost may be cheap.